### ISSUE TITLE
Custom IDs for Vectors & Index Partition Strategy

### ISSUE THREAD

**Issue by shriyog** [yahoojapan/NGT#105]
The search query result, which shows top N nearest vectors has `ID` field which seems to be the line number of vectors (ingested from a csv file). How can we specify a custom ID associated with vectors while ingestion? (can't find a way to provide `id` parameter in CLI, C++, or Go SDK either)

```
Query No.1
Rank	ID	Distance
1	3031	239.332
2	4079	240.002
3	3164	244.504
4	3718	246.763
5	157	  251.094
Query Time= 0.000472 (sec), 0.472 (msec)
```
It would be great if we could just attach identifiers while insertion & get back those in search results. Otherwise, we need to keep an external mapping between IDs in the business & NGT context.

If adding a custom id isn't supported what's the best way around it?

**Comment by masajiro** [comment]
You cannot specify your own custom IDs to your objects on NGT. It might be possible to implement insert functions with user's custom IDs to NGT, however, I have no time to update NGT at this moment. Another option is to use [Vald](https://github.com/vdaas/vald) that can manage user's metadata.

**Comment by shriyog** [comment]
Thanks a lot for clarifying. I understand & would keep an eye out for future release that supports custom IDs. For now, I'll work around by keeping a mapping to custom ids & line numbers i.e append position. And yes, we are exploring Vald too.

I had another question regarding the partitioning strategy for NGT index —  With a vector dataset of ~50 Mn, indexing it entirely on a single machine takes around 8 hrs (`ngt create -d 40 -D 2`). I was thinking of the following strategy
```
Index
- Split the vector dataset into N partitions & build index in parallel

Search
- Make search queries to all the partitions
- Merge the results
- Find top K final results with least distances
```

Will the results from an unpartitioned index be the same as the one with above partitioning strategy?

**Comment by masajiro** [comment]
Your strategy works well. Multiple indices might output more accurate results than a single index. Vald uses multiple distributed NGT indices like your strategy.

**Comment by shriyog** [comment]
@masajiro Thank you for acknowledging this. I tested recall rates on complete (46Mn vectors) vs partitioned index (6Mn vectors) & found accuracy increases as you mentioned. NGT's [eval utility](https://github.com/yahoojapan/NGT/issues/90#issuecomment-762496393) helped a lot. We should have added it to the Wiki or perhaps I'll raise a PR for the same.

Regarding Vald, we tried setting it up & benchmarking but it wasn't meeting perf expectations (possibly we missed something on the tuning part). We have a vector dataset of 60 Mn, 40 dimensions, & expect p99 search latency <10ms.

Currently, it seems using NGT library directly provides much more control, better performance, & flexibility. NGT decouples creating indexes from loading them, hence planning to leverage shared network volumes to isolate read & write.
